### PR TITLE
Add terraform mapping to k8s provider

### PR DESCRIPTION
**Pull request by Frassle**
### Proposed changes

    
This returns a basic mapping description for the terraform converter to map terraform programs using https://registry.terraform.io/providers/hashicorp/kubernetes to our kubernetes provider, even though it's not bridged.
    
 I grabbed a list of all terraform kubernetes resources by running `terraform providers schema -json` in a tiny terraform program that just defined a kubernetes provider. I've then hooked that up to gen-kubernetes to do a best guess of mapping everything (well just resources so far) from terraform to our provider. Most things look to map pretty well by convention, but we'll probably need some manual fixups to cover everything (already got one for the "container" to containers" rename).

### CODE REVIEW

File: provider/pkg/provider/provider.go [hunk 1]
```diff
@@ -210,8 +210,135 @@ func (k *kubeProvider) Call(ctx context.Context, req *pulumirpc.CallRequest) (*p
...
+		if putok == "" {
+			continue
+		}
+		resources[tftok] = map[string]interface{}{
+			"tok": putok,
+			"fields": map[string]interface{}{
+				"metadata": map[string]interface{}{
+					"maxItemsOne": true,
+				},
+				"spec": map[string]interface{}{
```
Comment (lblackstone) [inline provider/pkg/provider/provider.go:319]: `spec` isn't present on all k8s resources (`apiVersion`, `kind`, and `metadata` are the only guaranteed fields). I don't completely understand how this mapping is supposed to work, but I suspect this isn't right.

File: provider/pkg/provider/provider.go [hunk 2]
```diff
@@ -210,8 +210,135 @@ func (k *kubeProvider) Call(ctx context.Context, req *pulumirpc.CallRequest) (*p
...
+		"kubernetes_stateful_set":                        "kubernetes:apps/v1beta2:StatefulSet",
+		"kubernetes_stateful_set_v1":                     "kubernetes:apps/v1:StatefulSet",
+		"kubernetes_storage_class":                       "kubernetes:storage.k8s.io/v1beta1:StorageClass",
+		"kubernetes_storage_class_v1":                    "kubernetes:storage.k8s.io/v1:StorageClass",
+		"kubernetes_token_request_v1":                    "kubernetes:authentication.k8s.io/v1:TokenRequest",
+		"kubernetes_validating_webhook_configuration":    "kubernetes:admissionregistration.k8s.io/v1beta1:MutatingWebhookConfiguration",
+		"kubernetes_validating_webhook_configuration_v1": "kubernetes:admissionregistration.k8s.io/v1:MutatingWebhookConfiguration",
+	}
+
+	resources := make(map[string]interface{})
```
Comment (lblackstone) [inline provider/pkg/provider/provider.go:307]: nit: Use `any` rather than `interface{}`

File: provider/pkg/provider/provider.go (hunk 1 above)
Comment (Frassle) [reply provider/pkg/provider/provider.go:319]: 👍 I'm reworking this. I thought this simple map would suffice but turns out it's trickier (isn't it always 🙄)

**Comment by Frassle** [comment]
With this mapping in place the following terraform:
```
provider "kubernetes" {
}

resource "kubernetes_deployment" "nginx" {
  metadata {
    name = "scalable-nginx-example"
    labels = {
      App = "ScalableNginxExample"
    }
  }

  spec {
    replicas = 2
    selector {
      match_labels = {
        App = "ScalableNginxExample"
      }
    }
    template {
      metadata {
        labels = {
          App = "ScalableNginxExample"
        }
      }
      spec {
        container {
          image = "nginx:1.7.8"
          name  = "example"

          port {
            container_port = 80
          }

          resources {
            limits = {
              cpu    = "0.5"
              memory = "512Mi"
            }
            requests = {
              cpu    = "250m"
              memory = "50Mi"
            }
          }
        }
      }
    }
  }
}

resource "kubernetes_service" "nginx" {
  metadata {
    name = "nginx-example"
  }
  spec {
    selector = {
      App = kubernetes_deployment.nginx.spec.0.template.0.metadata[0].labels.App
    }
    port {
      node_port   = 30201
      port        = 80
      target_port = 80
    }

    type = "NodePort"
  }
}
```
Get converted to the following TypeScript:
```
import * as pulumi from "@pulumi/pulumi";
import * as kubernetes from "@pulumi/kubernetes";

const nginx = new kubernetes.apps.v1beta1.Deployment("nginx", {
    metadata: {
        name: "scalable-nginx-example",
        labels: {
            app: "ScalableNginxExample",
        },
    },
    spec: {
        replicas: 2,
        selector: {
            matchLabels: {
                app: "ScalableNginxExample",
            },
        },
        template: {
            metadata: {
                labels: {
                    app: "ScalableNginxExample",
                },
            },
            spec: {
                containers: [{
                    image: "nginx:1.7.8",
                    name: "example",
                    ports: [{
                        containerPort: 80,
                    }],
                    resources: {
                        limits: {
                            cpu: "0.5",
                            memory: "512Mi",
                        },
                        requests: {
                            cpu: "250m",
                            memory: "50Mi",
                        },
                    },
                }],
            },
        },
    },
});
const nginxResource = new kubernetes.core.v1.Service("nginxResource", {
    metadata: {
        name: "nginx-example",
    },
    spec: {
        selector: {
            app: nginx.spec.apply(spec => spec?.template?.metadata?.labels?.app),
        },
        ports: [{
            nodePort: 30201,
            port: 80,
            targetPort: 80,
        }],
        type: "NodePort",
    },
});
```

**Comment by lblackstone** [comment]
That conversion looks good except for the selected resource version. `v1beta1.Deployment` is deprecated and is only retained for backward compat on very old cluster versions.

**Comment by Frassle** [comment]
Ah cool I wasn't sure because TF had both kubernetes_deployment_v1 and kubernetes_deployment. I wasn't sure if the non versioned one should go to the first version released or the latest.

**Comment by lblackstone** [comment]
> Ah cool I wasn't sure because TF had both kubernetes_deployment_v1 and kubernetes_deployment. I wasn't sure if the non versioned one should go to the first version released or the latest.

I think picking the latest version should generally be safe/backward compatible. Lets do that for now, and can revise if we run into problems.

**Review by lblackstone — APPROVED** [review]
Changes LGTM, but is there any testing for this?

**Comment by Frassle** [comment]
> Changes LGTM, but is there any testing for this?

No, I wasn't sure what the best way to test this was?
We can write a .tf program and call convert on it, but should we convert to PCL and check the output of that or would y'all rather we pick a language (like typescript) and check it coverts all the way to that?

**Comment by lblackstone** [comment]
> No, I wasn't sure what the best way to test this was?

cc @t0yv0 @iwahbe @rquitales for thoughts

File: provider/cmd/pulumi-gen-kubernetes/main.go [hunk 3]
```diff
@@ -522,3 +527,228 @@ func mustWritePulumiSchema(pkgSpec schema.PackageSpec, version string) {
 	}
 	mustWriteFile(BaseDir, filepath.Join("sdk", "schema", "schema.json"), versionedSchemaJSON)
 }
+
+// Minimal types for reading the terraform schema file
+type TerraformAttributeSchema struct {
+}
```
Comment (iwahbe) [inline provider/cmd/pulumi-gen-kubernetes/main.go:533]:
```suggestion
type TerraformAttributeSchema struct {}
```

**Review by iwahbe — COMMENTED** [review]
YAML is fast, easy to target and aggressively type checked. HCL -> YAML and then previewing the yaml sounds like a great test.

**Comment by Frassle** [comment]
>  previewing the yaml sounds like a great test.

As in `pulumi preview`? Does that not need a k8s cluster spun up to run against?

**Comment by iwahbe** [comment]
> > previewing the yaml sounds like a great test.
> 
> As in `pulumi preview`? Does that not need a k8s cluster spun up to run against?

Why would `pulumi preview` require existing infrastructure? I'm not saying we should deploy anything, just use `pulumi preview` to perform a type check.

**Comment by Frassle** [comment]
> Why would pulumi preview require existing infrastructure?

Same reason an aws preview complains if your not yet logged into aws.

**Pull request merged**